### ISSUE TITLE
Global Environment Variables

### ISSUE THREAD

**Issue by x80486** [woodpecker-ci/woodpecker#936]
### Clear and concise description of the problem

I've notice that there is no way to declare a given environment variable to be global to all defined steps.

In `GitLab CI`, for instance, there is a "global" `variables` key that fulfills that purpose.

For instance, take a look at this manifest:

```yaml
default:
  cache:
    key: $CI_COMMIT_REF_SLUG
    paths:
      - $CI_PROJECT_DIR/.m2/repository/
  image: docker.io/library/eclipse-temurin:8-jdk-focal

stages:
  - compile

variables:
  MAVEN_CONFIG: --batch-mode --no-transfer-progress
  MAVEN_OPTS: -Dmaven.repo.local=$CI_PROJECT_DIR/.m2/repository/

compile:
  stage: compile
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_MERGE_REQUEST_ID
  script:
    - ./mvnw compile $MAVEN_EXTRA_OPTS
  variables:
    MAVEN_EXTRA_OPTS: -DskipTests=true
```

Here `MAVEN_CONFIG` and `MAVEN_OPTS` apply to all steps. As far as I can tell, currently there is no way to accomplish this in `Woodpecker CI`.

### Suggested solution

An `environment` key at the root level.

> The `environment` key at the step level would have to account for overrides, in case the same environment variable is declared in both.

### Alternative

N/A

### Additional context

N/A

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by 6543** [comment]
the yaml spec already allow us to do this: https://yaml.org/spec/1.1/#id863390

https://blog.jasonhee.com/how-to-use-yaml-aliases

we could document this ...

**Comment by 6543** [comment]
example: https://codeberg.org/6543/ci-service-example/pulls/1/files

**Comment by 6543** [comment]
well one goal is to have a clean & simple pipeline config as mouch as posible, and if it's an official spec feature why should we reinvent the wheel?

only one thing we might have to do: ajust the lint step -> json schema and mention a "space" where you could define them
at the moment woodpecker-cli lint will error but with woodpecker-server it just works

tldr: let this issue be open until we have a nice docu about it and the linter do understand it

**Comment by svdHero** [comment]
@6543 I use this simple `.woodpecker.yml` file with Woodpecker 0.15 and Gitea 1.17.3:

```
variables:
  &set-global-env-vars
    - export MY_FOO="hello"
    - export MY_BAR="world"

pipeline:
  fist-stage:
    image: alpine:latest
    commands:
      - *set-global-env-vars
      - env | sort
```
However, when I push the repo to Gitea, the pipeline is not triggered and all I get is a 400 Webhook response in Gitea saying:
```
failure to parse yaml from hook
```

What's wrong here? I also tried [this approach](https://stackoverflow.com/a/72366027/3760986) using
```
.set-global-env-vars: &set-global-env-vars
    - export MY_FOO="hello"
    - export MY_BAR="world"

pipeline:
  fist-stage:
    image: alpine:latest
    commands:
      - *set-global-env-vars
      - env | sort
```
but with the same result.

Any advice?

**Comment by waldner** [comment]
YAML aliases do not work when trying to embed the referenced anchor into another string, the way true env variables would.
